PMD validate error

**Describe the bug**
When the Azure Devops pipeline runs PMD static code analysis via sfpowerscripts sfpowerkit as per the following command line, 
npx sfdx sfpowerkit:source:pmd -d  force-app-package_src_delta/force-app -f  text -r  setup/pmd-ruleset.xml --version=6.36.0 --loglevel INFO

I gives the following warning
{"name":"Pmd","hostname":"fv-az69-443","pid":2528,"level":40,"msg":"[\n  'WARNING:',\n  '--ruleset has been deprecated and will be removed in a future version. Use --rulesets instead.'\n]","time":"2021-06-28T09:50:29.900Z","v":0}

WARNING: --ruleset has been deprecated and will be removed in a future version. Use --rulesets instead.

and the tasks exist with error code 4
##[error]Exit with error code 4

**To Reproduce**
1. sfdx sfpowerkit:source:pmd -d  force-app-package_src_delta/force-app -f  text -r  setup/pmd-ruleset.xml --version=6.36.0 --loglevel INFO

**Expected behavior**
None of the errors highlighted where critical errors so the task should have passed.

**Screenshots**
![image](https://user-images.githubusercontent.com/56622820/123619010-0b404a00-d809-11eb-8902-615cdfb5de37.png)
![image](https://user-images.githubusercontent.com/56622820/123620340-66bf0780-d80a-11eb-9684-e0f6d9699b2d.png)
![image](https://user-images.githubusercontent.com/56622820/123620430-822a1280-d80a-11eb-8c56-73077ed0f590.png)

**Desktop (please complete the following information):**

Installing plugin sfpowerkit... yarn add v1.22.10
Installing plugin sfpowerkit... info No lockfile found.Installing plugin sfpowerkit... [1/4] Resolving packages...Installing plugin sfpowerkit... [2/4] Fetching packages...Installing plugin sfpowerkit... [3/4] Linking dependencies...Installing plugin sfpowerkit... [4/4] Building fresh packages...Installing plugin sfpowerkit... success Saved lockfile.Installing plugin sfpowerkit... success Saved 339 new dependencies.Installing plugin sfpowerkit... info Direct dependenciesInstalling plugin sfpowerkit... └─ sfpowerkit@3.1.5Installing plugin sfpowerkit... info All dependenciesInstalling plugin sfpowerkit... ├─ @hapi/bourne@2.0.0Installing plugin sfpowerkit... ├─ @salesforce/command@3.1.3Installing plugin sfpowerkit... ├─ @types/sinon@10.0.2Installing plugin sfpowerkit... ├─ async-listener@0.6.10Installing plugin sfpowerkit... ├─ big-integer@1.6.48Installing plugin sfpowerkit... ├─ buffer-indexof-polyfill@1.0.2Installing plugin sfpowerkit... ├─ clone@2.1.2Installing plugin sfpowerkit... ├─ crc32-stream@4.0.2Installing plugin sfpowerkit... ├─ diagnostic-channel-publishers@0.4.4Installing plugin sfpowerkit... ├─ emoji-regex@8.0.0Installing plugin sfpowerkit... ├─ fast-json-stable-stringify@2.1.0Installing plugin sfpowerkit... ├─ fs.realpath@1.0.0Installing plugin sfpowerkit... ├─ heroku-client@3.1.0Installing plugin sfpowerkit... ├─ is-glob@4.0.1Installing plugin sfpowerkit... ├─ jmespath@0.15.0Installing plugin sfpowerkit... ├─ jwa@1.4.1Installing plugin sfpowerkit... ├─ lodash.get@4.4.2Installing plugin sfpowerkit... ├─ minimist@1.2.5Installing plugin sfpowerkit... ├─ nice-try@1.0.5Installing plugin sfpowerkit... ├─ optional-js@1.3.1Installing plugin sfpowerkit... ├─ path-key@2.0.1Installing plugin sfpowerkit... ├─ propagate@2.0.1Installing plugin sfpowerkit... ├─ readdir-glob@1.1.1Installing plugin sfpowerkit... ├─ reusify@1.0.4Installing plugin sfpowerkit... ├─ shebang-regex@1.0.0Installing plugin sfpowerkit... ├─ sprintf-js@1.0.3Installing plugin sfpowerkit... ├─ tar-stream@2.2.0Installing plugin sfpowerkit... ├─ type-detect@4.0.8Installing plugin sfpowerkit... ├─ walkdir@0.0.11Installing plugin sfpowerkit... ├─ xtend@4.0.2Installing plugin sfpowerkit... └─ zip-stream@4.1.0Installing plugin sfpowerkit... Done in 13.01s.Installing plugin sfpowerkit... installed v3.1.5
warning sfpowerkit > request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
warning sfpowerkit > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning sfpowerkit > salesforce-alm > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning sfpowerkit > salesforce-alm > request-promise-native@1.0.9: request-promise-native has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
warning sfpowerkit > @salesforce/core > jsforce > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning sfpowerkit > request > har-validator@5.1.5: this library is no longer supported
warning sfpowerkit > request > uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
warning sfpowerkit > salesforce-alm > @salesforce/source-deploy-retrieve > @salesforce/core > jsforce > request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
warning "sfpowerkit > ts-node@9.1.1" has unmet peer dependency "typescript@>=2.7".
warning "sfpowerkit > @salesforce/command > @oclif/command@1.8.0" has unmet peer dependency "@oclif/config@^1".
The plugin [sfpowerkit] is not digitally signed but it is allow-listed.
Finished digital signature check.

--- a/packages/core/src/command/SFDXCommand.ts
+++ b/packages/core/src/command/SFDXCommand.ts
@@ -1,5 +1,5 @@
 import ExecuteCommand from "./commandExecutor/ExecuteCommand";
-import { Logger, LoggerLevel } from "../logger/SFPLogger";
+import SFPLogger, { Logger, LoggerLevel } from "../logger/SFPLogger";
 
 export abstract class SFDXCommand {
   public constructor(
@@ -14,6 +14,7 @@ export abstract class SFDXCommand {
     if (quiet) command += ` --json`;
     command += " " + this.getGeneratedParams();
 
+    SFPLogger.log("Generated Command:"+command,LoggerLevel.TRACE,this.logger);
     let executor: ExecuteCommand = new ExecuteCommand(this.logger,this.logLevel);
     let output = await executor.execCommand(command, this.project_directory);
     if (quiet) {

--- a/packages/core/src/sfpowerkitwrappers/AnalyzeWithPMDImpl.ts
+++ b/packages/core/src/sfpowerkitwrappers/AnalyzeWithPMDImpl.ts
@@ -3,16 +3,15 @@ import SFPLogger, { Logger, LoggerLevel } from "../logger/SFPLogger";
 
 export default class AnalyzeWithPMDImpl extends SFDXCommand {
   public constructor(
-    protected project_directory: string,
-    private directory: string,
-    private ruleset: string,
+    protected sourceDirectory: string,
+    private ruleset: string, //Future Use
     private format: string,
     private ouputPath: string,
     private version: string,
     protected logger?: Logger,
     protected logLevel?: LoggerLevel
   ) {
-    super(null, project_directory,logger,logLevel);
+    super(null, sourceDirectory,logger,logLevel);
   }
 
   getSFDXCommand(): string {
@@ -25,7 +24,7 @@ export default class AnalyzeWithPMDImpl extends SFDXCommand {
     let command:string="";
 
 
-    if (this.directory) command=`-d  ${this.directory}`;
+    if (this.sourceDirectory) command=`-d  ${this.sourceDirectory}`;
     if (this.format)    command +=` -f  ${this.format}`;
     if (this.ouputPath) command+=` -o  ${this.ouputPath}`;
     if (this.version)   command+=` --version  ${this.version}`;

--- a/packages/core/src/sfpowerkitwrappers/AnalyzeWithPMDImpl.ts
+++ b/packages/core/src/sfpowerkitwrappers/AnalyzeWithPMDImpl.ts
@@ -1,53 +1,37 @@
-import child_process = require("child_process");
-import { onExit } from "../utils/OnExit";
-import { isNullOrUndefined } from "util";
-import SFPLogger from "../logger/SFPLogger";
+import { SFDXCommand } from "../command/SFDXCommand";
+import SFPLogger, { Logger, LoggerLevel } from "../logger/SFPLogger";
 
-
-export default class AnalyzeWithPMDImpl {
-
-
-  public constructor(private project_directory:string, private directory: string, private ruleset:string, private format:string, private ouputPath: string, private version:string) {}
-
-  public async exec(command: string): Promise<void> {
-
-    let child=child_process.exec(command,  { encoding: "utf8",maxBuffer: 1024 * 1024*5, cwd:this.project_directory });
-
-    child.stdout.on("data",data=>{SFPLogger.log(data.toString()); });
-    child.stderr.on("data",data=>{SFPLogger.log(data.toString()); });
-
-
-    await onExit(child);
-
+export default class AnalyzeWithPMDImpl extends SFDXCommand {
+  public constructor(
+    protected project_directory: string,
+    private directory: string,
+    private ruleset: string,
+    private format: string,
+    private ouputPath: string,
+    private version: string,
+    protected logger?: Logger,
+    protected logLevel?: LoggerLevel
+  ) {
+    super(null, project_directory,logger,logLevel);
   }
 
-  public async buildExecCommand(): Promise<string> {
+  getSFDXCommand(): string {
+    return "sfdx sfpowerkit:source:pmd";
+  }
+  getCommandName(): string {
+    return "PMD Analysis";
+  }
+  getGeneratedParams(): string {
+    let command:string="";
 
-    let command;
-        command = `sfdx sfpowerkit:source:pmd`;
 
-
-    if(!isNullOrUndefined(this.directory))
-    command+=` -d  ${this.directory}`;
-
-
-    if(!isNullOrUndefined(this.format))
-    command+=` -f  ${this.format}`;
-
-    if(!isNullOrUndefined(this.ouputPath))
-    command+=` -o  ${this.ouputPath}`;
-
-    if(!isNullOrUndefined(this.ruleset) && this.ruleset.length>0)
-    command+=` -r  ${this.ruleset}`;
-
-    if(!isNullOrUndefined(this.version))
-    command+=` --version=${this.version}`;
-
-    command+=` --loglevel INFO`
-
-    SFPLogger.log(command);
+    if (this.directory) command=`-d  ${this.directory}`;
+    if (this.format)    command +=` -f  ${this.format}`;
+    if (this.ouputPath) command+=` -o  ${this.ouputPath}`;
+    if (this.version)   command+=` --version  ${this.version}`;
+    command +=` --loglevel ${LoggerLevel[SFPLogger.logLevel]}`;
+    
+    
     return command;
   }
-
-
 }

--- a/packages/sfpowerscripts-cli/src/commands/sfpowerscripts/analyze/pmd.ts
+++ b/packages/sfpowerscripts-cli/src/commands/sfpowerscripts/analyze/pmd.ts
@@ -1,29 +1,29 @@
-import { flags } from '@salesforce/command';
-import SfpowerscriptsCommand from '../../../SfpowerscriptsCommand';
-import { Messages, SfdxError } from '@salesforce/core';
-import AnalyzeWithPMDImpl from '@dxatscale/sfpowerscripts.core/lib/sfpowerkitwrappers/AnalyzeWithPMDImpl';
-import xml2js = require('xml2js');
-import {isNullOrUndefined} from 'util';
-const fs = require('fs');
-const path = require('path');
-const os = require('os');
+import { flags } from "@salesforce/command";
+import SfpowerscriptsCommand from "../../../SfpowerscriptsCommand";
+import { Messages } from "@salesforce/core";
+import AnalyzeWithPMDImpl from "@dxatscale/sfpowerscripts.core/lib/sfpowerkitwrappers/AnalyzeWithPMDImpl";
+import xml2js = require("xml2js");
+import SFPLogger, { ConsoleLogger } from "@dxatscale/sfpowerscripts.core/lib/logger/SFPLogger";
+const fs = require("fs");
 
 // Initialize Messages with the current plugin directory
 Messages.importMessagesDirectory(__dirname);
 
 // Load the specific messages for this file. Messages from @salesforce/command, @salesforce/core,
 // or any library that is using the messages framework can also be loaded this way.
-const messages = Messages.loadMessages('@dxatscale/sfpowerscripts', 'analyze_with_PMD');
+const messages = Messages.loadMessages(
+  "@dxatscale/sfpowerscripts",
+  "analyze_with_PMD"
+);
 
 export default class AnalyzeWithPMD extends SfpowerscriptsCommand {
-
-  public static description = messages.getMessage('commandDescription');
+  public static description = messages.getMessage("commandDescription");
 
   public static examples = [
     `$ sfdx sfpowerscripts:analyze:pmd -b\n`,
     `Output variable:`,
     `sfpowerscripts_pmd_output_path`,
-    `<refname>_sfpowerscripts_pmd_output_path`
+    `<refname>_sfpowerscripts_pmd_output_path`,
   ];
 
   protected static requiresProject = true;
@@ -31,14 +31,55 @@ export default class AnalyzeWithPMD extends SfpowerscriptsCommand {
   protected static requiresDevhubUsername = false;
 
   protected static flagsConfig = {
-    sourcedir: flags.string({description: messages.getMessage('sourceDirectoryFlagDescription')}),
-    ruleset: flags.string({description: messages.getMessage('rulesetFlagDescription'), options: ['sfpowerkit','Custom'], default: 'sfpowerkit'}),
-    rulesetpath: flags.string({description: messages.getMessage('rulesetPathFlagDescription')}),
-    format: flags.string({description: messages.getMessage('formatFlagDescription'), options: ['text','textcolor','csv','emacs','summaryhtml','html','xml','xslt','yahtml','vbhtml','textpad', 'sarif'], default: 'text'}),
-    outputpath: flags.string({char: 'o', description: messages.getMessage('outputPathFlagDescription')}),
-    version: flags.string({description: messages.getMessage('versionFlagDescription'), default: '6.26.0'}),
-    istobreakbuild: flags.boolean({char: 'b', description: messages.getMessage('isToBreakBuildFlagDescription')}),
-    refname: flags.string({description: messages.getMessage('refNameFlagDescription')}),
+    sourcedir: flags.string({
+      description: messages.getMessage("sourceDirectoryFlagDescription"),
+    }),
+    ruleset: flags.string({
+      description: messages.getMessage("rulesetFlagDescription"),
+      options: ["sfpowerkit", "Custom"],
+      default: "sfpowerkit",
+    }),
+    rulesetpath: flags.string({
+      description: messages.getMessage("rulesetPathFlagDescription"),
+    }),
+    format: flags.string({
+      description: messages.getMessage("formatFlagDescription"),
+      options: [
+        "text",
+        "textcolor",
+        "csv",
+        "emacs",
+        "summaryhtml",
+        "html",
+        "xml",
+        "xslt",
+        "yahtml",
+        "vbhtml",
+        "textpad",
+        "sarif",
+      ],
+      default: "text",
+    }),
+    outputpath: flags.string({
+      char: "o",
+      description: messages.getMessage("outputPathFlagDescription"),
+    }),
+    version: flags.string({
+      required: false,
+      default: "6.34.0",
+      description: messages.getMessage("versionFlagDescription"),
+    }),
+    istobreakbuild: flags.boolean({
+      char: "b",
+      deprecated: {
+        messageOverride:
+          "--istobreakbuild has been deprecated, the command will always break",
+      },
+      description: messages.getMessage("isToBreakBuildFlagDescription"),
+    }),
+    refname: flags.string({
+      description: messages.getMessage("refNameFlagDescription"),
+    }),
     loglevel: flags.enum({
       description: "logging level for this command invocation",
       default: "info",
@@ -57,34 +98,22 @@ export default class AnalyzeWithPMD extends SfpowerscriptsCommand {
         "ERROR",
         "FATAL",
       ],
-    })
+    }),
   };
 
-
-  public async execute(){
+  public async execute() {
     try {
-
-      console.log("Test.. PMD");
-
       const source_directory: string = this.flags.sourcedir;
       const ruleset: string = this.flags.ruleset;
 
-
-      let rulesetpath=""
+      let rulesetpath = "";
       if (ruleset == "Custom") {
-        let rulesetpath = this.flags.rulesetpath;
-        console.log(rulesetpath);
+        console.log("Custom Rulesets are temporaily disabled, Please check back in few days");
       }
-
-
 
       const format: string = this.flags.format;
       const outputPath: string = this.flags.outputpath;
       const version: string = this.flags.version;
-
-      const isToBreakBuild = this.flags.istobreakbuild;
-
-      let result: [number, number, number] = [0, 0, 0];
 
       let pmdImpl: AnalyzeWithPMDImpl = new AnalyzeWithPMDImpl(
         null,
@@ -92,81 +121,57 @@ export default class AnalyzeWithPMD extends SfpowerscriptsCommand {
         rulesetpath,
         format,
         outputPath,
-        version
-      );
-      let command = await pmdImpl.buildExecCommand();
-      await pmdImpl.exec(command);
-
-      if (!isNullOrUndefined(this.flags.refname)) {
-        if (!isNullOrUndefined(outputPath)) {
-            fs.writeFileSync('.env', `${this.flags.refname}_sfpowerscripts_pmd_output_path=${outputPath}\n`, {flag:'a'});
-        } else {
-            fs.writeFileSync('.env', `${this.flags.refname}_sfpowerscripts_pmd_output_path=${process.env.PWD}/pmd-output\n`, {flag:'a'});
-        }
-      } else {
-        if (!isNullOrUndefined(outputPath)) {
-          fs.writeFileSync('.env', `sfpowerscripts_pmd_output_path=${outputPath}\n`, {flag:'a'});
-        } else {
-          fs.writeFileSync('.env', `sfpowerscripts_pmd_output_path=${process.env.PWD}/pmd-output\n`, {flag:'a'});
-        }
-      }
-
-      let artifactFilePath = path.join(
-      os.homedir(),
-      "sfpowerkit",
-      "pmd",
-      `pmd-bin-${version}`,
-      "sf-pmd-output.xml"
+        version,
+        new ConsoleLogger(),
+        SFPLogger.logLevel
       );
 
-      if (fs.existsSync(artifactFilePath)) {
-      result = parseXmlReport(artifactFilePath);
-      }
+      await pmdImpl.exec(false);
 
-      if (isToBreakBuild && result[2] > 0)
-          throw new SfdxError(`Build Failed due to ${result[2]} critical defects found`);
+     
 
     } catch (err) {
       console.log(err);
       // Fail the task when an error occurs
       process.exit(1);
     }
+  }
 
-    function parseXmlReport(xmlReport: string): [number, number, number] {
-      let fileCount = 0;
-      let violationCount = 0;
-      let criticaldefects = 0;
+  private parseXmlReport(xmlReport: string): [number, number, number] {
+    let fileCount = 0;
+    let violationCount = 0;
+    let criticaldefects = 0;
 
-      let reportContent: string = fs.readFileSync(xmlReport, "utf-8");
-      xml2js.parseString(reportContent, (err, data) => {
-        // If the file is not XML, or is not from PMD, return immediately
-        if (!data || !data.pmd) {
-          console.debug(`Empty or unrecognized PMD xml report ${xmlReport}`);
-          return null;
-        }
+    let reportContent: string = fs.readFileSync(xmlReport, "utf-8");
+    xml2js.parseString(reportContent, (err, data) => {
+      // If the file is not XML, or is not from PMD, return immediately
+      if (!data || !data.pmd) {
+        console.debug(`Empty or unrecognized PMD xml report ${xmlReport}`);
+        return null;
+      }
 
-        if (!data.pmd.file || data.pmd.file.length === 0) {
-          // No files with violations, return now that it has been marked for upload
-          return null;
-        }
+      if (!data.pmd.file || data.pmd.file.length === 0) {
+        // No files with violations, return now that it has been marked for upload
+        return null;
+      }
 
-        data.pmd.file.forEach((file: any) => {
-          if (file.violation) {
-            fileCount++;
-            violationCount += file.violation.length;
-          }
-        });
-
-        for (let i = 0; i < data.pmd.file.length; i++) {
-          data.pmd.file[i].violation.forEach(element => {
-            if (element["$"]["priority"] == 1) {
-              criticaldefects++;
-            }
-          });
+      data.pmd.file.forEach((file: any) => {
+        if (file.violation) {
+          fileCount++;
+          violationCount += file.violation.length;
         }
       });
 
-      return [violationCount, fileCount, criticaldefects];
-    }
+      for (let i = 0; i < data.pmd.file.length; i++) {
+        data.pmd.file[i].violation.forEach((element) => {
+          if (element["$"]["priority"] == 1) {
+            criticaldefects++;
+          }
+        });
+      }
+    });
+
+    return [violationCount, fileCount, criticaldefects];
   }
+
 }

--- a/packages/sfpowerscripts-cli/src/commands/sfpowerscripts/analyze/pmd.ts
+++ b/packages/sfpowerscripts-cli/src/commands/sfpowerscripts/analyze/pmd.ts
@@ -116,7 +116,6 @@ export default class AnalyzeWithPMD extends SfpowerscriptsCommand {
       const version: string = this.flags.version;
 
       let pmdImpl: AnalyzeWithPMDImpl = new AnalyzeWithPMDImpl(
-        null,
         source_directory,
         rulesetpath,
         format,

